[UnifiedPDF] Selections are broken in rotated pages

aprotyas opened the pull request:
#### 099c347a91fc9f6cc329b487dde7d154968dd188
<pre>
[UnifiedPDF] Selections are broken in rotated pages
<a href="https://bugs.webkit.org/show_bug.cgi?id=270073">https://bugs.webkit.org/show_bug.cgi?id=270073</a>
<a href="https://rdar.apple.com/123621642">rdar://123621642</a>

Reviewed by Simon Fraser.

The `-[PDFPage drawWithBox:toContext:]` method accounts for page
rotation, and we took this behavior to assume that the PDFSelection
drawing API would behave similarly. Unfortunately, this is wrong, and
said method expects callers to apply the transform corresponding to
the appropriate boundary box type.

This patch amends this situation by manually applying the transform in
question to the context&apos;s CTM while we have a selection painted into
that context.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.h:

Expose `toPageTransform` and `geometryForPage` publicly.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.mm:
(WebKit::PDFDocumentLayout::geometryForPage const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::paintPDFContent):

Canonical link: <a href="https://commits.webkit.org/275351@main">https://commits.webkit.org/275351@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/69e524535d647f780bea1c444e203a1eb60c1dba

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41603 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20617 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43983 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44172 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37695 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23729 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17947 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42177 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17538 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35834 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/15055 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/15247 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/45581 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37784 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37164 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40917 "layout-tests (failure)") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/16418 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13481 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/39328 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18037 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9321 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18093 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/17681 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -748,8 +748,12 @@ static String mutationObserverNotificationString()
 
-        if (haveSelection)
+        if (haveSelection) {
+            AffineTransform transformForBox { [page transformForBox:kPDFDisplayBoxCropBox] };
```
Should we get this transform from `PDFDocumentLayout` (akin to `PDFDocumentLayout::toPageTransform()`) instead?

aprotyas replied:
Indeed this is just the inverse of `toPageTransform()`.

aprotyas replied:
Using that instead in https://github.com/WebKit/WebKit/commit/69e524535d647f780bea1c444e203a1eb60c1dba

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -748,8 +748,12 @@ static String mutationObserverNotificationString()
+            AffineTransform transformForBox { [page transformForBox:kPDFDisplayBoxCropBox] };
+            context.concatCTM(transformForBox);
             [m_currentSelection drawForPage:page.get() withBox:kCGPDFCropBox active:isVisibleAndActive inContext:context.platformContext()];
+            context.concatCTM(transformForBox.inverse().value_or(AffineTransform { }));
```
Nice to just use a `GraphicsContextStateSaver`.

aprotyas replied:
Yes, I was looking for something like this. I'll use it.

aprotyas replied:
https://github.com/WebKit/WebKit/commit/69e524535d647f780bea1c444e203a1eb60c1dba

aprotyas committed:
[UnifiedPDF] Selections are broken in rotated pages
https://bugs.webkit.org/show_bug.cgi?id=270073
rdar://123621642

Reviewed by Simon Fraser.

The `-[PDFPage drawWithBox:toContext:]` method accounts for page
rotation, and we took this behavior to assume that the PDFSelection
drawing API would behave similarly. Unfortunately, this is wrong, and
said method expects callers to apply the transform corresponding to
the appropriate boundary box type.

This patch amends this situation by manually applying the transform in
question to the context's CTM while we have a selection painted into
that context.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.h:

Expose `toPageTransform` and `geometryForPage` publicly.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.mm:
(WebKit::PDFDocumentLayout::geometryForPage const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::paintPDFContent):

Canonical link: https://commits.webkit.org/275351@main
